Core hole vasp

ndaelman-hu opened the pull request:
Identification of the core-hole (location + orbital level) with appropriate references to `run.system.atoms_group` so topology view is leveraged. This will be further extended to FHIaims and ATOMPAW in separate PRs.

ndaelman-hu committed:
- Add `parse_core_hole` and link to `run.system`
- TODO: Correct mapping to `CoreHole.atom_index`, now it's just species index

ndaelman-hu committed:
- Update mapping to new schema
- Add check for presence flag

ndaelman-hu committed:
- Set defaults for core index, n, l, occ
- Make mapping global

ndaelman-hu committed:
Change name of `simulation.method.CoreHole` to `CoreHoleSpectra`

ndaelman-hu committed:
Remove unfinished class in QE

ndaelman-hu committed:
- Update wiring between `CoreHole` and `AtomsGroup`
- TODO: add lower atom layer

ndaelman-hu committed:
- Update parser mapping of `CoreHole`, `AtomParameters`, and `System`
- TODO: Make sure `AtomParameters.core_hole` also displays on the GUI side, not just the archive

ndaelman-hu committed:
Remove debug logging for `atom_parameters.core_hole`

ndaelman-hu committed:
Fix improper reference at `core_hole.atomsgroup_ref`

ndaelman-hu commented:
Left to do:
- [x] Apply ICORELEVEL tag at the electron count level
- [x] Check linting
- [ ] Add core-hole test (with varying occupations or multiple sites?)
- [ ] Add documentation?

JosePizarro3 commented on electronicparsers/vasp/parser.py:
```diff
@@ -1366,7 +1371,58 @@ def parse_kpoints(self, section):
+            type='core_hole',
+            atom_indices=atom_ids,
+            n_atoms=len(atom_ids),
+            atoms_group=sub_groups,
```
Yes, this is the wrong point I am mentioning in the Gitlab. I will do a more thorough review, but initially I'd say that `atoms_group` is an unecesserary repetition, as you are creating this extra layer.

You only need to store something like:
```
atoms_group = AtomsGroup(
   type='core_hole',
   label=elem_label,
   n_atoms=len(atom_ids),
   atom_indices=atom_ids,
)
```

Anything else (`label='core-hole_' + elem_label` and `atoms_group=sub_groups`) is the repetition I am mentioning that you don't need 🙂

ndaelman-hu replied:
> initially I'd say that atoms_group is an unecesserary repetition

Point taken. Let's keep this discussion over at Gitlab, and if we decide against this layer, I'll simply remove it.

ndaelman-hu replied:
> Anything else (label='core-hole_' + elem_label and atoms_group=sub_groups) is the repetition I am mentioning that you don't need

Yeah, the current `label` was just a placeholder, so feel free to raise suggestions.
I applied your code, but this means that the term _core-hole_ isn't visible anymore at the 2nd visual layer (fig. 1).
If we go this way, we really do need a 3rd visual layer, just to denote the type. Else, this information should be in the title / label.

Something similar happens with Projection, where the word doesn't show up at all (fig. 2)... Granted, the calculation is labeled as a Wannier projection, so the user can likely guess what the sublayer means. Note that **this does not apply to a calculation with a core-hole**: it's just a regular `SinglePoint`, `GeometryOptimization`, etc.
I'd keep it that way.

Btw, the **all-caps is a nightmare for elemental labels**. In the examples, it's just TI or CU, respectively. I'd like to amend that together with Lauri.

![Screenshot from 2023-11-17 20-15-39](https://github.com/nomad-coe/electronic-parsers/assets/107392603/c2e2dd6c-a403-4c83-9302-b6123ac95345)

![Screenshot from 2023-11-17 20-17-41](https://github.com/nomad-coe/electronic-parsers/assets/107392603/99ed6a67-56c9-44a4-b0aa-2dac2873563a)

JosePizarro3 replied:
My main concern here is that: 3 layers is unnecessary, as we only need (1) the original crystal + (2) the specific `active_atoms` used and their environments. 

Now I see that the naming for `label` might be confusing. We can then go back to your case (`label=f'core_hole_{elem_id}'`) or define in gitlab a new structural type called `'active_atoms'` (or `['core_hole_atoms', 'projection_atoms']`. I'd personally prefer the first option, but maybe the second has its benefits

ndaelman-hu committed:
- Add support for systems with core-holes at multiple sites
- Rearrange formal connection between higher and lower `atoms_group`

JosePizarro3 commented on electronicparsers/vasp/parser.py:
```diff
@@ -1434,7 +1489,11 @@ def parse_method(self):
                     sec_hubb.x_vasp_projection_type = 'on-site'
             atom_counts[element[i]] += 1
-        sec_method.x_vasp_atom_kind_refs = sec_method.atom_parameters
+        sec_method.x_vasp_atom_kind_refs = sec_method.atom_parameters  # TODO: deprecate x_vasp_atom_kind_refs?
```
Yes, please. Deprecate it.

JosePizarro3 commented on electronicparsers/vasp/parser.py:
```diff
@@ -1547,7 +1610,7 @@ def parse_workflow(self):
         self.archive.workflow2 = workflow
 
-    def parse_configurations(self):
+    def parse_configurations(self, atoms_group: Union[AtomsGroup, None] = None):
```
Is this the correct typing? Shouldn't be `atoms_group: AtomsGroup = None`?

ndaelman-hu replied:
It can be simplified, yes, but not to your suggestion. The type `AtomsGroup` only includes the class, not `None` objects.

I will change it to `def parse_configurations(self, atoms_group: Optional[None] = None):`

JosePizarro3 replied:
That's a weird typing (Optional None to None?), but nvm, this is then deprecated if you pass `method_dict: dict = {}` 😄

ndaelman-hu replied:
Oops, that's a typo. It has to be `def parse_configurations(self, atoms_group: Optional[AtomsGroup] = None)`, obviously.

JosePizarro3 commented on electronicparsers/vasp/parser.py:
```diff
@@ -1834,8 +1904,11 @@ def parse(self, filepath, archive, logger):
 
-        self.parse_configurations()
+        # Note: the logic for a proper topology might have to become more complex
+        self.parse_configurations(
```
It seems to me that `self.parse_configurations` will be more readable if using `method_dict` as an input, and then within the function you resolve `atoms_group`.

Just a small style detail.

ndaelman-hu replied:
Sounds fine. I'll just add an explanation on the exact meaning and use of `method_dict`.

ndaelman-hu replied:
I held off on resolving this discussions due to changes at the time.
No arguments are being passed to `parse_configurations`, so it is essentially taken care off.

Just a small point of feedback. While your suggestion definitely improves readability, it comes with the downside of coupling the input of `parse_configurations` to the output from `parse_method`. That coupling is too strong, and therefore a fatal design flaw.
In this case, it would be better to keep this format (closer to functional programming), OR have `parse_configurations` collect the various output packages of other methods `parse_...`.
In that case it's still coupled, but less. The final decision depends on how likely those methods are to be removed.

JosePizarro3 reviewed (commented):
Very nice. Please, don't forget to change also `workflowparsers/quantum_espresso_xspectra` with the new `CoreHoleSpectra`.

I will keep up with the discussion about nesting AtomsGroup.

**Btw, can you share some examples and add testing?**

ondracka commented:
I can't comment much on the VASP parser changes, but rather on the general metainfo changes (doing this here due to no access to the upstream github). In general this would suit my needs, but at least for Wien2k for states with n>1 where a spin-orbit splitting happens, like 2p states, I can actually specify if I want the corehole for 2p1/2 or 2p3/2, and equivalently for d and f states. I could likely also specify up and down spin state with Wien2k (but I never did that).
I checked the VASP docs and there is seems this is not possible, but I would expect that maybe other all-electron codes can do this as well.
My another favorite code OpenMX also allows all kinds of core-holes (there is nice table at https://www.openmx-square.org/openmx_man3.9/node192.html) there it also depends whether one has collinear or non-collinear spin polarization. So maybe we need more quantum numbers...

ondracka commented:
One question is if we also want to somehow explicitly include somewhere the charge info at this point. Together with with the corehole usually also comes the question about where to put the extra charge. I.e., if the electron was removed from the cell (and a homogeneous background change was added so that the calculations behave in a sane way) but it is actually a charged cell (this is usually done for insulators) or if the electron was placed at the Fermi level (this would make sense for core-hole in metals). This is not related just to the core-hole and is relevant also for other methods (for charged-defect calculations in insulators, etc.). I believe VASP always adds the background charge automatically for example (and it will probably diverge for a true charged cell anyway).)

ndaelman-hu commented:
> I can't comment much on the VASP parser changes, but rather on the general metainfo changes (doing this here due to no access to the upstream github). In general this would suit my needs, but at least for Wien2k for states with n>1 where a spin-orbit splitting happens, like 2p states, I can actually specify if I want the corehole for 2p1/2 or 2p3/2, and equivalently for d and f states. I could likely also specify up and down spin state with Wien2k (but I never did that). I checked the VASP docs and there is seems this is not possible, but I would expect that maybe other all-electron codes can do this as well. My another favorite code OpenMX also allows all kinds of core-holes (there is nice table at https://www.openmx-square.org/openmx_man3.9/node192.html) there it also depends whether one has collinear or non-collinear spin polarization. So maybe we need more quantum numbers...

Thank you for pointing that out! I'll open up tickets for OpenMX then, and merge the schema after having working parsers for both parsers.

ndaelman-hu commented:
> One question is if we also want to somehow explicitly include somewhere the charge info at this point. Together with with the corehole usually also comes the question about where to put the extra charge. I.e., if the electron was removed from the cell (and a homogeneous background change was added so that the calculations behave in a sane way) but it is actually a charged cell (this is usually done for insulators) or if the electron was placed at the Fermi level (this would make sense for core-hole in metals). This is not related just to the core-hole and is relevant also for other methods (for charged-defect calculations in insulators, etc.). I believe VASP always adds the background charge automatically for example (and it will probably diverge for a true charged cell anyway).)

Very good point! We keep track of the no. (valence) electrons and overall system charge under `run.method.electronic`. We could just improve the normalizers here and make this information more visible in the Entry Overview. This would allow for finding core-holes with the electron exited to the Fermi level or the vacuum.

But maybe we want something **more explicit** when core-holes are being used? VASP also allows partial core-holes, as well as initial and final state approximations. @ondracka Are these also used in Wien2k or OpenMX? That would help me decide on their placement in the schema.

ondracka commented:
> Thank you for pointing that out! I'll open up tickets for OpenMX then, and merge the schema after having working parsers for both parsers.

Let me know if you need some test-cases. Here is a simple C 1s corehole in OpenMX: https://nomad-lab.eu/prod/v1/staging/gui/search/calculations/entry/id/zLqikxv-mfCrI3sC9EuXmqfEX04T but I can also produce something simpler and more suitable for eventual inclusion in parser unit tests (or also for the other core-holes than 1s). For Wien2k I have some N1s core-hole: https://nomad-lab.eu/prod/v1/staging/gui/search/entries/entry/id/Fbk7yucgh7CxMsp0RKdR263J_OIj , O1s https://nomad-lab.eu/prod/v1/staging/gui/search/entries/entry/id/1GH5Rvm-t2tAYPrL9UdtkShaas3s, or Ti2p3/2 https://nomad-lab.eu/prod/v1/staging/gui/search/entries/entry/id/s5W753xr30EulmEUzsu75HMtxZMh coreholes in NOMAD already.

ondracka commented:
> Very good point! We keep track of the no. (valence) electrons and overall system charge under `run.method.electronic`. We could just improve the normalizers here and make this information more visible in the Entry Overview. This would allow for finding core-holes with the electron exited to the Fermi level or the vacuum.
> 
> But maybe we want something **more explicit** when core-holes are being used? VASP also allows partial core-holes, as well as initial and final state approximations. @ondracka Are these also used in Wien2k or OpenMX? That would help me decide on their placement in the schema.

Partial core-holes with arbitrary fraction are easy in Wien2k (and probably other LAPW all electron codes), however I'm not aware that this would be possible in OpenMX.
Regarding the "initial" states, i.e., I assume you are referring to what VASP would be doing with ICORELEVEL=1 this is what all-electron codes do by default (and at that point there is no core-hole at all actually, just VASP includes the core-levels in the calculations instead of using pseudopotential), so for all-electron codes this is just standard groundstate. For OpenMX "initial state" one actually needs specific pseudopotential and basis set for the atom with the core-hole that include the core-level in question, but no other setting are needed. So I think that for VASP run with ICORELEVEL=1 there would be no CoreHole section and the fact that the extra core electron are explicitly included would be encoded somehow in the pseudopotential section.
"Final state" is the true core-hole calculation.

ndaelman-hu committed:
- Remove lowest visualization layer with independent atoms
- Change labels

ndaelman-hu committed:
Convert `projection` type to `active_atom`

ndaelman-hu committed:
Update element label to `projection` in Wannier

ndaelman-hu committed:
Fix labelling of highlighted atom site

ndaelman-hu committed:
- Apply schema changes to `CoreHole`
- Remove redundant `elem_label` from `parse_corehole`

ndaelman-hu committed:
- Add extraction of system charge and valence electrons
- TODO: improve PP extraction from vasprun.xml files

ndaelman-hu committed:
- Change naming `AtomicState` to `AtomicOrbitalState`
- Apply reviewer's feedback

ndaelman-hu committed:
- Correct core-hole occupation
- Fix system electron count and charge

ndaelman-hu committed:
Update CLZ default

ndaelman-hu committed:
Rename `active_atom` to `active_orbital`

ndaelman-hu committed:
- Add core-level method (IS, FS) used
- Handle cases without `atom_parameter.n_electrons`

ndaelman-hu committed:
Remove redundant `x_vasp_atom_kind_refs`

ndaelman-hu commented:
> but at least for Wien2k for states with n>1 where a spin-orbit splitting happens, like 2p states, I can actually specify if I want the corehole for 2p1/2 or 2p3/2, and equivalently for d and f states. I could likely also specify up and down spin state with Wien2k (but I never did that).

This is supported now.

> My another favorite code OpenMX also allows all kinds of core-holes (there is nice table at https://www.openmx-square.org/openmx_man3.9/node192.html) there it also depends whether one has collinear or non-collinear spin polarization. So maybe we need more quantum numbers...

Let me look into this. Maybe support for non-collinear calculations requires its own PR. Feel free to make any suggestions.

ndaelman-hu commented:
> Let me know if you need some test-cases. Here is a simple C 1s corehole in OpenMX: https://nomad-lab.eu/prod/v1/staging/gui/search/calculations/entry/id/zLqikxv-mfCrI3sC9EuXmqfEX04T but I can also produce something simpler and more suitable for eventual inclusion in parser unit tests (or also for the other core-holes than 1s).

@ondracka It seems that OpenMX reads the pseudopotentials from a different folder. Since the files are not copied over to the calculation folder, I guess it's unlikely that people will upload them.

Personally, I prefer reading the original pseudopotential file, but I guess that here it's better to build out an interpreter of the filename. Do you agree? How reliable is this approach?

ndaelman-hu commented:
> Regarding the "initial" states, i.e., I assume you are referring to what VASP would be doing with ICORELEVEL=1

Indeed, this is the terminology they use in the [VASP Wiki](https://www.vasp.at/wiki/index.php/ICORELEVEL).

> this is what all-electron codes do by default (and at that point there is no core-hole at all actually, just VASP includes the core-levels in the calculations instead of using pseudopotential), so for all-electron codes this is just standard groundstate.

@ondracka I fully agree with your explanation of the methodology.
However, in most contexts, ICORELEVEL=1is employed with Koopmans theorem or Slater's Transition method to produce CLBEs. One such example (not with VASP) would be [here](https://pubs.aip.org/aip/jcp/article/158/9/094111/2881303/Slater-transition-methods-for-core-level-electron). An overview of the various approaches is given in the first 2 paragraphs on page 2.

So, I see 2 options:
1. I infer that anyone using I**CORELEVEL**=1, does so with the intention of computing a core-level. There are several ways to approximate this (Koopman, or any of the fractional occupations listed above). If I know which state is targeted, I could even produce all the numbers.
2. I leave it up to the user to specify if they used the orbitals for computing the core-hole, as well as the actual scheme they used. This would come closer to what you proposed.

Let me know what you think.

ondracka commented:
> Let me look into this. Maybe support for non-collinear calculations requires its own PR. Feel free to make any suggestions.

I think this is OK for now, lets revisit the non-collinear case when we actually have some use cases...

ondracka commented:
> Personally, I prefer reading the original pseudopotential file, but I guess that here it's better to build out an interpreter of the filename. Do you agree? How reliable is this approach?

I'm only aware of the one set of pseudos provided by the OpenMX developers (versioned by the releases), so it should be quite reliable (unless people really handcraft their own pseudos, but that should be not too common).

ndaelman-hu committed:
- Update interfacing `parse_core_hole` with the rest of the parser
- Change behaviour `ICORELEVEL=1`, which is no longer associated with core-holes necessarily, but rather relaxed core states
- Remove `x_vasp_core_hole_method`

ondracka commented:
I went quickly through the actual code changes and while I was not reviewing the logic per se, the general idea how things are done now looks fine, so I'm happy with how this turned out in general.

ndaelman-hu committed:
Add units for `charge` under `run`

ndaelman-hu committed:
- Add `parse_species` to the OpenMX parser, mapping the species definition onto the pseudopotentials and core holes
- Update `pseudpotential.type` in the VASP parser to reflect US Vanderbilt

ndaelman-hu committed:
- Store output `parse_species` into the archive
- Correct `parse_species` logic

ndaelman-hu committed:
- Fix core_hole storage
- Add labe to `atom_parameters`

ndaelman-hu commented:
@ladinesa @lauri-codes Design question: core-holes are kind of connected to various other sections (too much interconnectivity):
1. Each `run.method.CoreHole` has to reference an `run.system.AtomsGroup` (this also applies to other sections, such as `CoreHoleSpectra`). Atm, I set these manually in the parsers, but I get the feeling that this could be done by a normalizer. I guess one could loop over all `AtomParameters` which contain `CoreHole`, but it sounds suboptimal: it requires the atom indexes too... Moreover, they might be other sections that do not fall under `AtomsParameters`. I'd rather have a push-system that works for any `AtomsGroup` - Methods.subs_section pair. Any ideas?
2. The computation of `Method.charge` is contingent on `CoreHole`. Same remark as above.
3. `CoreHole` is in many cases specified in the same parsed blocks / files as `Pseudopotential`. I'm trying to package both together in each new implementation now.

ndaelman-hu committed:
Connect `CoreHole` to `AtomsGroup`

lauri-codes commented:
I think the parser is ultimately responsible for setting all connections in place. A normalizer could very well be used to e.g. check the validity of the data or to add new derived data.

Note that it is of course fine to have a single unified **function** in e.g. `nomad` or in the `electronicparsers` package that is shared between all parsers and resolves all the references. I think this approach is often used in the `atomisticparsers` package, where several tasks are common for the parsers.

ndaelman-hu commented:
> I think the parser is ultimately responsible for setting all connections in place. A normalizer could very well be used to e.g. check the validity of the data or to add new derived data.
> 
> Note that it is of course fine to have a single unified **function** in e.g. `nomad` or in the `electronicparsers` package that is shared between all parsers and resolves all the references. I think this approach is often used in the `atomisticparsers` package, where several tasks are common for the parsers.

I get your point, in that the logical structure of the output may vary and this is where the parser has to jump in, of course.
However, the schema on our side will always be `AtomsGroup` pointing to `CoreHole`, where the latter may alter the overall `charge` (and is often set along with the `Pseudopotential`).

For anyone extending the parsers, this structure is far from obvious, nor is it clearly communicated anywhere.
My first instinct, for example, was to extend `AtomsGroup`, rather than `Method`. I think we have to communicate these kinds of designs better to other (external) devs, since we're headed in the direction of a modular platform.

Maybe a _normalizer_ is not appropriate here. Another option that come to mind is a `FactoryMethod`, which delivers (maybe even sets) a `AtomsGroup` - `CoreHole` pair. If we go this route, where would this best be added? In `util.py`?

lauri-codes commented:
My first suggestion would be to include the method in the `electronicparsers` repo if it is only relevant for these parsers. But if the implementation should be shared across different parser repos, then `util.py` is an option.

ndaelman-hu commented:
@ondracka From the [OpenMX documentation](https://www.openmx-square.org/openmx_man3.9/node192.html), I'm understanding that CLBEs are computed using $\Delta$-SCF between the _initial_ (containing the corrected pseudopotential) and _final_ state (corrected pseudopotential + electron removed). Is this correct?

I'm asking, since the terminology is confusing with [VASP's](https://www.vasp.at/wiki/index.php/ICORELEVEL), where _initial-state approximation_ (`icorelevel = 1`) is a single SCF with Koopman. There, the $\Delta$-SCF is between a regular groundstate (same pseudopotential) and a _final-state approximation_ (`icorelevel = 2`).

ondracka commented:
> @ondracka From the [OpenMX documentation](https://www.openmx-square.org/openmx_man3.9/node192.html), I'm understanding that CLBEs are computed using Δ-SCF between the _initial_ (containing the corrected pseudopotential) and _final_ state (corrected pseudopotential + electron removed). Is this correct?

That is my understanding as well.

ndaelman-hu commented:
> > @ondracka From the [OpenMX documentation](https://www.openmx-square.org/openmx_man3.9/node192.html), I'm understanding that CLBEs are computed using Δ-SCF between the _initial_ (containing the corrected pseudopotential) and _final_ state (corrected pseudopotential + electron removed). Is this correct?
> 
> That is my understanding as well.

Okay, great! I'm close to finishing with OpenMX too, will request you to review then.

ndaelman-hu committed:
Add extraction of `core.hole` tag

ndaelman-hu committed:
- Flesh out business logic in `OpenMX` parser
- Rewrite business logic to use `n_excited_electrons` in the `VASP` parser

ndaelman-hu committed:
Optimize projector logic in `Wannier90` parser

ndaelman-hu committed:
Add nc support for `OpenMX`

ndaelman-hu committed:
Update vasp and openmx parsers after overhaul

ndaelman-hu committed:
Change to label `active_orbitals`

ndaelman-hu committed:
Update Wannier90 tests

JosePizarro3 commented:
Btw, is something included in the TBStudio parser regarding the active atom?

ndaelman-hu committed:
Update Wannier90 imports after rebasing

ndaelman-hu committed:
Rectify mistake after rebasing

ndaelman-hu committed:
- Fix typing issues
- Adopt `mypy` configurations from main project

ndaelman-hu commented:
> Btw, is something included in the TBStudio parser regarding the active atom?

Hey, thx for mentioning TBStudio. I had to rebase first, now I can see that parser too.
Atm, it does not assign any `atoms_group`s, so it won't break with any of my changes.
I will add support when extending `core_hole` to atomic orbitals in general. I'll open an issue this evening.

ndaelman-hu committed:
Ratify electron count from elements in VASP

ladinesa commented on electronicparsers/openmx/parser.py:
```diff
@@ -244,6 +286,91 @@ class OpenmxParser:
     def __init__(self):
         pass
 
+    @property
```
why is this a  property?

ndaelman-hu replied:
Simply easier access: essentially this is a dynamically generated dictionary
It's accessed 2x later on

ladinesa replied:
why not simply have a get function. a property should depend on a state somehow.

ndaelman-hu replied:
I'm sorry, I'm not seeing the distinction: is the `atom_index_dict` a state of `self.archive`?
I can also change this to a `cached_property` decorator, but that might be overkill for 2-time use

ndaelman-hu replied:
Sorry, it's actually accessed in a loop, so caching might be advantageous

ladinesa replied:
but everytime you access this, it evaluates it. you can have self._atom_index_dict that can be reset if some dependency changes.

ndaelman-hu replied:
> but everytime you access this, it evaluates it

as is, yes, but not if I cache it

ladinesa commented on electronicparsers/openmx/parser.py:
```diff
@@ -244,6 +286,91 @@ class OpenmxParser:
+            definitions[1] = remove_extension(definitions[1])
+            definitions[2] = remove_extension(definitions[2])
+        except IndexError:
+            logger.error(f'Species definition must be of length 3: {definitions}')
```
should you return here?

ladinesa commented on electronicparsers/openmx/parser.py:
```diff
@@ -244,6 +286,91 @@ class OpenmxParser:
+                    n_electrons_excited=1,
+                )
+            except KeyError:
+                logger.error(f'Unknown l-quantum symbol: {quantum_nums[1]}')
```
should you return here?

ladinesa commented on electronicparsers/openmx/parser.py:
```diff
@@ -414,7 +545,7 @@ def parse(self, mainfile: str, archive: EntryArchive, logger):
         self.parse_workflow()
 
-        self.parse_method()
+        self.parse_method(logger)
```
no self.logger?

ndaelman-hu replied:
Is it not passed as an argument of `parse`?

ladinesa replied:
what i meant was you do not need to pass as arg to parse_method if it is class attr.

ndaelman-hu replied:
Is it a class attribute? I'm only seeing it passed as an argument to `OpenmxParser.parse()`.
Moreover, wouldn't the alternative be more of an object attribute instead?

ladinesa replied:
Yes that was I was asking if it is a class attr, if not maybe add it under parse.

ndaelman-hu replied:
I was following the example of `parse_structure`, but this is a free-standing method.
We might want to stratify the interface.

ladinesa replied:
yes this is already implemented in the abstract class for parsers that I am developing in https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-FAIR/-/merge_requests/1538

ladinesa commented on electronicparsers/openmx/parser.py:
```diff
@@ -471,4 +602,23 @@ def parse(self, mainfile: str, archive: EntryArchive, logger):
                             sec_calc.time = time * units.fs
 
+        # set core-hole atoms groups
+        sec_system = self.archive.run[-1].system[-1]
```
error handling here

ladinesa commented on electronicparsers/vasp/parser.py:
```diff
@@ -1497,6 +1531,26 @@ def parse_method(self):
+            except AttributeError:
+                break
+            try:
+                species_electrons -= param.core_hole.n_electrons_excited
```
what is the initial value for species_electrons

ndaelman-hu replied:
Isn't it `param.n_electrons`? Else the loop breaks.

ladinesa replied:
i see

ndaelman-hu replied:
can I resolve?

ladinesa commented on electronicparsers/vasp/parser.py:
```diff
@@ -1575,6 +1629,20 @@ def parse_system(n_calc):
+            ## note that this check should be added to `parse_core_hole` if other kinds of atom_parameters are set
+            if self.parser.incar.get('ICORELEVEL', 0) == 2:
+                core_hole, core_hole_group, corehole_id = self.parse_core_hole()
+                self.archive.run[-1].method[-1].atom_parameters[corehole_id].core_hole = core_hole
```
no need for error handling here?

ladinesa reviewed (commented):
Some miinor suggestions.

ndaelman-hu commented on electronicparsers/vasp/parser.py:
```diff
@@ -1575,6 +1629,20 @@ def parse_system(n_calc):
+                core_hole, core_hole_group, corehole_id = self.parse_core_hole()
+                self.archive.run[-1].method[-1].atom_parameters[corehole_id].core_hole = core_hole
+
+                if sec_system.atoms_group is None:
+                    sec_system.atoms_group = []
```
@ladinesa I guess this is redundant then, thanks to the instantiation process

ladinesa replied:
yes it is. I think this is still a remnant of the old infra where the default for repeating sections is None.

ndaelman-hu committed:
Apply reviewer feedback

ndaelman-hu commented:
> Very nice. Please, don't forget to change also `workflowparsers/quantum_espresso_xspectra` with the new `CoreHoleSpectra`.
> 
> I will keep up with the discussion about nesting AtomsGroup.
> 
> **Btw, can you share some examples and add testing?**

@JosePizarro3 Can you pls lift the change request in your review. That applied to an old version.
Your suggestions have been incorporated.
